File: examples3d/newton_cradle3.rs
use rapier_testbed3d::Testbed;
use rapier3d::prelude::*;

pub fn init_world(testbed: &mut Testbed) {
    /*
     * World
     */
    let mut world = PhysicsWorld::new();

    let radius = 0.5;
    let length = 10.0 * radius;
    let rb = RigidBodyBuilder::dynamic();
    let co = ColliderBuilder::ball(radius).restitution(1.0);

    let n = 5;

    for i in 0..n {
        let (ball_pos, attach) = (
            Vector::new(i as Real * 2.2 * radius, 0.0, 0.0),
            Vector::Y * length,
        );
        let vel = if i >= n - 1 {
            Vector::new(7.0, 0.0, 0.0)
        } else {
            Vector::ZERO
        };

        let ground = world
            .bodies
            .insert(RigidBodyBuilder::fixed().translation(ball_pos + attach));
        let rb = rb.clone().translation(ball_pos).linvel(vel);
        let (handle, _) = world.insert(rb, co.clone());

        let joint = SphericalJointBuilder::new().local_anchor2(attach);
        world.insert_impulse_joint(ground, handle, joint);
    }

    /*
     * Set up the testbed.
     */
    testbed.set_physics_world(world);
    testbed.look_at(Vec3::new(10.0, 10.0, 10.0), Vec3::ZERO);
}

File: python/rapier-testbed/rapier_testbed/examples3/newton_cradle3.py
"""Port of examples3d/newton_cradle3.rs."""
from __future__ import annotations

import rapier3d as rp
from .._registry import register

CATEGORY = "Collisions"
NAME = "Newton cradle"


def init_world(testbed) -> None:
    bodies = rp.RigidBodySet()
    colliders = rp.ColliderSet()
    impulse_joints = rp.ImpulseJointSet()
    multibody_joints = rp.MultibodyJointSet()

    radius = 0.5
    length = 10.0 * radius
    n = 5

    for i in range(n):
        ball_pos = (i * 2.2 * radius, 0.0, 0.0)
        attach = (0.0, length, 0.0)
        if i >= n - 1:
            vel = (7.0, 0.0, 0.0)
        else:
            vel = (0.0, 0.0, 0.0)

        ground = bodies.insert(
            rp.RigidBody.fixed().translation(
                (ball_pos[0] + attach[0], ball_pos[1] + attach[1], ball_pos[2] + attach[2])
            )
        )
        body = rp.RigidBody.dynamic().translation(ball_pos).linvel(vel)
        h = bodies.insert(body)
        colliders.insert_with_parent(rp.Collider.ball(radius).restitution(1.0), h, bodies)

        joint = (
            rp.SphericalJoint.builder()
            .local_anchor2(attach)
            .build()
        )
        impulse_joints.insert(ground, h, joint, wake_up=True)

    testbed.set_world(bodies, colliders, impulse_joints, multibody_joints)
    testbed.look_at((10.0, 10.0, 10.0), (0.0, 0.0, 0.0))


register(CATEGORY, NAME, init_world, dim=3)


if __name__ == "__main__":
    from .._picker import run

    run(initial=f"{CATEGORY} / {NAME}")
